### PR TITLE
feat: aliyun direct mail connector

### PR DESCRIPTION
**Pull request by wangsijie**
## Summary

Add the first connector: Aliyun DirectMail.

Notion document: https://www.notion.so/silverhand/Aliyun-DirectMail-5d80077184d44ff38ffc164e6ba957c6

## Testing

Tested in local machine with real credentials and received the exptected email.

`index.ts` does not have a test file, it is a bit complicated, leave it to LOG-1046.

### CODE REVIEW

File: packages/core/src/connectors/aliyun-dm/index.ts
Comment (IceHe) [inline packages/core/src/connectors/aliyun-dm/index.ts:22]: `AliyunDmConfig` : `M` → `m` better camel-case naming?

File: packages/core/src/connectors/types.ts
Comment (gao-sun) [inline packages/core/src/connectors/types.ts:18]: we can just use PascalCase for constant keys, i.e. `SignIn`

File: packages/core/src/connectors/aliyun-dm/index.ts
Comment (gao-sun) [reply packages/core/src/connectors/aliyun-dm/index.ts:17]: should we use `keyof EmailMessageTypes`

File: packages/core/src/connectors/aliyun-dm/request.ts
Comment (wangsijie) [reply packages/core/src/connectors/aliyun-dm/request.ts:60]: with `& PublicParameters`, IDE will suggest predefined keys in `PublicParameters`

File: packages/core/src/connectors/aliyun-dm/index.ts
Comment (wangsijie) [reply packages/core/src/connectors/aliyun-dm/index.ts:51]: `@doc` added to function and interface.

File: packages/core/src/connectors/aliyun-dm/request.ts
Comment (gao-sun) [reply packages/core/src/connectors/aliyun-dm/request.ts:63]: good point. in this case, actually we need `& { [key: string]?: string }`?

Comment (wangsijie) [reply packages/core/src/connectors/aliyun-dm/request.ts]: Lint suggest use `Record` instead of `{ [key: string]?: string }`

File: packages/core/src/connectors/aliyun-dm/send-single-mail.ts
Comment (simeng-li) [reply packages/core/src/connectors/aliyun-dm/send-single-mail.ts:7]: The doc link provided above? Since there isn't a SDK to use, all the request types are hardcoded.

File: packages/core/src/connectors/aliyun-dm/send-single-mail.ts
Comment (simeng-li) [inline packages/core/src/connectors/aliyun-dm/send-single-mail.ts:29]: could define as a const

File: packages/core/src/connectors/aliyun-dm/index.ts
Comment (simeng-li) [inline packages/core/src/connectors/aliyun-dm/index.ts:12]: nit: maybe put it into a const config file, along with the connector domain name?

File: packages/core/src/connectors/aliyun-dm/index.ts
Comment (simeng-li) [inline packages/core/src/connectors/aliyun-dm/index.ts:40]: wondering, since we passed in `metadata.type` as an input param to the `getConnectorConfig` util method. Should we keep the template filter by type logic within the method itself?  Instead of returning all the templates?

File: packages/core/src/connectors/utilities/aliyun.ts
Comment (simeng-li) [inline packages/core/src/connectors/utilities/aliyun.ts:67]: nit: since the code is straightforward enough, maybe keeping it inline is good enough? No need to declare new variables.

File: packages/core/src/connectors/utilities/aliyun.ts
Comment (simeng-li) [inline packages/core/src/connectors/utilities/aliyun.ts:54]: similarly, maybe we could have a const config file for these hardcoded congfigs.

File: packages/core/src/connectors/utilities/aliyun.ts
Comment (simeng-li) [inline packages/core/src/connectors/utilities/aliyun.ts:77]: Object.entries()?

File: packages/core/src/connectors/aliyun-dm/index.ts
Comment (wangsijie) [reply packages/core/src/connectors/aliyun-dm/index.ts:12]: The outside don't need to know the `domain name`. Metadata here is meant for necessary info for Logto core.

File: packages/core/src/connectors/utilities/aliyun.ts
Comment (wangsijie) [reply packages/core/src/connectors/utilities/aliyun.ts:77]: Lint rules suggest using `for...in`

File: packages/core/src/connectors/utilities/aliyun.ts
Comment (wangsijie) [reply packages/core/src/connectors/utilities/aliyun.ts:54]: `aliyun` is a single util file, let's keep the config inside the file until `aliyun` becomes a directory when necessary in the future?

Comment (simeng-li) [reply packages/core/src/connectors/utilities/aliyun.ts:54]: o I was thinking of a config file for all connectors.  Any way your call, not a blocker at all.

Comment (wangsijie) [reply packages/core/src/connectors/utilities/aliyun.ts:54]: each connector is designed to be "independant", so keep it inside each dir.

File: packages/core/src/connectors/aliyun-dm/index.ts
Comment (wangsijie) [reply packages/core/src/connectors/aliyun-dm/index.ts:46]: Actually, "templates" is this specific connector's private stuff, contrubitors may provider another `aliyun-dm` connector without this "templates" design.
So is a bit different from `getConnectorConfig`. `getConnectorConfig` is a common util for all connectors.

File: packages/core/src/connectors/utilities/aliyun.ts
Comment (darcyYe) [inline packages/core/src/connectors/utilities/aliyun.ts:18]: `+` -> `%20`?

Comment (darcyYe) [reply packages/core/src/connectors/utilities/aliyun.ts:18]: image

**Review by darcyYe — CHANGES_REQUESTED** [review]

File: packages/core/src/connectors/utilities/aliyun.ts
Comment (darcyYe) [inline packages/core/src/connectors/utilities/aliyun.ts:25]: nit: use `canonicalizedQuery` to make it more clear?
```suggestion
  const canonicalizedQuery = Object.keys(parameters)
```

File: packages/core/src/connectors/utilities/aliyun.ts
Comment (darcyYe) [inline packages/core/src/connectors/utilities/aliyun.ts:38]: same here
```suggestion
  const stringToSign = `${method.toUpperCase()}&${escaper('/')}&${escaper(canonicalizedQuery)}`;
```

File: packages/core/src/connectors/aliyun-dm/index.ts
Comment (gao-sun) [inline packages/core/src/connectors/aliyun-dm/index.ts:44]: since now we only use `id` for identifying, should the function reflects this change?

File: packages/core/src/connectors/aliyun-dm/index.ts
Comment (gao-sun) [inline packages/core/src/connectors/aliyun-dm/index.ts:49]: should this be an enum with i18n key? since the error looks general.

File: packages/core/src/connectors/aliyun-dm/index.ts
Comment (gao-sun) [inline packages/core/src/connectors/aliyun-dm/index.ts:61]: node 16 supports `replaceAll` already?

File: packages/core/src/connectors/aliyun-dm/send-single-mail.ts
Comment (gao-sun) [inline packages/core/src/connectors/aliyun-dm/send-single-mail.ts:28]: `sendSingleMail` & `SingleSendMail` are confusing. would you like to align to usage?

**Review by gao-sun — APPROVED** [review]
approve to unblock. overall lgtm.

File: packages/core/src/connectors/aliyun-dm/send-single-mail.ts
Comment (darcyYe) [reply packages/core/src/connectors/aliyun-dm/send-single-mail.ts:28]: this naming convention comes from aliyun DM docs lol

File: packages/core/src/connectors/aliyun-dm/index.ts
Comment (darcyYe) [reply packages/core/src/connectors/aliyun-dm/index.ts:44]: I fixed this in my PR, maybe ignore this.

**Review by darcyYe — APPROVED** [review]
overall LGTM

File: packages/core/src/connectors/aliyun-dm/index.ts
Comment (wangsijie) [reply packages/core/src/connectors/aliyun-dm/index.ts:44]: Thanks.

File: packages/core/src/connectors/aliyun-dm/send-single-mail.ts
Comment (wangsijie) [reply packages/core/src/connectors/aliyun-dm/send-single-mail.ts:28]: It sucks, but still aligned with `singleSendMail`

File: packages/core/src/connectors/aliyun-dm/index.ts
Comment (wangsijie) [reply packages/core/src/connectors/aliyun-dm/index.ts:61]: Can not find document for now, but YES, tested.

File: packages/core/src/connectors/aliyun-dm/index.ts
Comment (wangsijie) [reply packages/core/src/connectors/aliyun-dm/index.ts:49]: LOG-1335

**Pull request merged**
